src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -222,6 +228,78 @@ public static RequestDelegateResult Create(MethodInfo methodInfo, Func<HttpConte
         return HandleRequestBodyAndCompileRequestDelegate(responseWritingMethodCall, factoryContext);
     }
 
+    private static void AddTypeProvidedMetadata(MethodInfo methodInfo, FactoryContext factoryContext)
+    {
+        const string argumentNullExceptionMessage = "The IEnumerable<object> returned from GetMetadata must not be null.";
+
+        // Get metadata from parameter types
+        var parameters = methodInfo.GetParameters();
+        foreach (var parameter in parameters)
+        {
+            if (typeof(IProvideEndpointParameterMetadata).IsAssignableFrom(parameter.ParameterType))
+            {
+                // Parameter type implements IProvideEndpointParameterMetadata
+                var metadata = GetMetadataForParameterMethod.MakeGenericMethod(parameter.ParameterType).Invoke(null, new object?[] { parameter, factoryContext.ServiceProvider });
+                if (metadata is null)
+                {
+                    throw new ArgumentNullException(null, argumentNullExceptionMessage);
+                }
+
+                factoryContext.Metadata.AddRange((IEnumerable<object>)metadata);
+            }
+
+            if (typeof(IProvideEndpointMetadata).IsAssignableFrom(parameter.ParameterType))
+            {
+                // Parameter type implements IProvideEndpointMetadata
+                var metadata = GetMetadataForEndpointMethod.MakeGenericMethod(parameter.ParameterType).Invoke(null, new object?[] { methodInfo, factoryContext.ServiceProvider });
+                if (metadata is null)
+                {
+                    throw new ArgumentNullException(null, argumentNullExceptionMessage);
+                }
+
+                factoryContext.Metadata.AddRange((IEnumerable<object>)metadata);
+            }
+        }
+
+        // Get metadata from return type
+        if (methodInfo.ReturnType is not null && typeof(IProvideEndpointMetadata).IsAssignableFrom(methodInfo.ReturnType))
+        {
+            // Return type implements IProvideEndpointMetadata
+            var metadata = GetMetadataForEndpointMethod.MakeGenericMethod(methodInfo.ReturnType).Invoke(null, new object?[] { methodInfo, factoryContext.ServiceProvider });
+            if (metadata is null)
+            {
+                throw new ArgumentNullException(null, argumentNullExceptionMessage);
+            }
+
+            factoryContext.Metadata.AddRange((IEnumerable<object>)metadata);
+        }
+
+        // Get metadata from method attributes that implement IProvideEndpointMetadata
+        var methodAttrs = methodInfo.GetCustomAttributes(inherit: true).OfType<IProvideEndpointMetadata>();
+        foreach (var attribute in methodAttrs)
+        {
+            var metadata = GetMetadataForEndpointMethod.MakeGenericMethod(attribute.GetType()).Invoke(null, new object?[] { methodInfo, factoryContext.ServiceProvider });
```

Comment:
Wait, the attribute can implement the interface as well to provide metadata? This makes me wonder if we need both attributes but I guess they are different scenarios.

Reply:
Isn't this kind of redundant since the attributes already are metadata? Is this just a way for a single attribute to supply multiple pieces of metadata?

Should we be doing this same thing for parameter attributes? That seems more useful since those aren't already added as metadata themselves.